### DISCUSSION TITLE
The database is not seeded until admin "/_" page is accessed

### DISCUSSION THREAD

**Discussion by dschissler** [pocketbase/pocketbase#3845] [Q&A]
```golang
	if app.Settings().Meta.AppUrl != url {
		form := forms.NewSettingsUpsert(app)
		form.Meta.AppUrl = url
		return nil, errors.New("Unable to set Meta.AppURL")
	}
```

I have this in an `OnAfterBootstrap` event.  It fails upon first run with error `no such table: _params`.

I think that these tables should be built upon first init without needing user involvement.  It makes initial application setup difficult to reconcile with normal use.

So maybe two steps are required such as initing admin records.  But really that should be able to happen after this.  I'm just trying to build something now (on a Raspberry Pi for the first time ever), and this slowed me down a lot.

**Comment by ganigeorgiev** [comment]
Unapplied migrations (including system ones) are automatically executed only when `serve` is used, aka. attach to the `OnBeforeServe` hook and the above should work just fine.

Alternatively, you can move the initial settings creation in a migration - https://pocketbase.io/docs/go-migrations/#initialize-default-application-settings.

`OnAfterBootstrap` hook is triggered after the db connections and the minimum app initialization has been completed.

**Reply by dschissler** [reply]
Are migrations global?

**Reply by ganigeorgiev** [reply]
The default migration runner uses `migrations.AppMigrations` to load the migrations from, if that's what you are asking (_option to configure it through the app config struct exists in my todo and will be added eventually in the future_).

**Reply by dschissler** [reply]
It seems like that it registers globally.

```golang
    m "github.com/pocketbase/pocketbase/migrations"

    m.Register(func(db dbx.Builder) error {
```

**Reply by dschissler** [reply]
Is there a reason why migrations run only within `Serve`?

In generally I'd like to see it be possible to control the normal building of the database for different execution modes.  In my main professional project I'm using Pocketbase in two modes.  One is a `charmbracelt/wish` SSH server that is used for RPC.  I use the Pocketbase database in it.  It's great because Dao provides everything that I need.  It's a massive time saver compared to setting up a MySQL database.

I faced some difficulty setting it up in the development environment because it also needs to run in the same Golang executable as my  Svelte front end website.  After some finagling I managed to combine the two.

The problem is that Pocketbase is highly tied to the API being the only way to run it, but it's also a perfectly fine backup backend database ~~system~~ app.

Another issue was attempting to create a command for installation of a systemd file in a custom Cobra command.  The problem is that Pocketbase starting is too tightly coupled to database directory being created, but in this instance I want to run my "PB as framework" executable as root without a data directory.  The point is just to write the correct systemd files and to run the right commands to install it.  It's not seeming to be possible right now, and so I had to push those steps further back to the provision system.

Then again right now I want to set the App Name on a fresh install (as I'm just trying to get everything to compile) and it informs me that the tables haven't been created.

Basically I want to use Pocketbase as my goto framework for all manners of applications, but it's still too inflexible for that to be ergonomic.  I really like what you did by refactoring a lot of the serve command stuff into the api struct, but for me the cleanup  still could go a bit further.

Again, the idea for me is that in the most generic sense, I only need a RootCmd and a database.  From there the Web API is available, and I've built up a rather complex system on top of that with Casbin, templates, Vite manifest integration, etc, etc.  But I don't always want that for every little program and use case.  With just a little refactoring it could be possible.

**Reply by dschissler** [reply]
@ganigeorgiev My thinking is like this.  If I needed to create a small Golang CLI utility or long-running process then I would need to start somewhere.  I'd need Cobra to be able to have CLI options.  At this point already it just makes sense to use Pocketbase because it also comes with a really nice database.  Even if I'm not yet sure if the tool needs to store data then it already is perhaps even less work to use than to set it all up myself.

Then later if a health check single access point needs to be made then a potential webserver is already available.  That is a huge time savings.  Just starting with Pocketbase is the way to do it, unless there is a specific reason why it shouldn't be.

The current problem is that Pocketbase isn't currently designed to be used without using any of its features.  You've spent so much time making great features, but now I think that you need to allow us to more easily not use them.

In particular around the creation of the data directory unless `-h`, `--help` is found.  That makes it hard to build in a final installation step like installing a systemd service file, etc.  It's a weird problem that makes it hard to use a Pocketbase program as a more general utility.

Also the bootstrap and init related events really assume that you're going to be using the web API.  As I just learned, migrations are tied to the api.  It's also necessary to run the api blocking server last and before others.  Ideally all generic system level events would happen first, and then only api specific migrations would happen when the api starts.

------------------------

p.s. When I have free days I'm working on PB on a RPI 4B.  I'm developing directly on the device using [openvscode-server](https://github.com/gitpod-io/openvscode-server), which works very well.  The problem is Golang recompile times though.  So the obvious solution is to use both Golang and Javascript framework modes.  So in a sense I'm using it as intended, but with a custom version of Pocketbase.

I found a nice simple http multi-part pseudo camera stream project (which actually uses far less precious RPI CPU than a normal stream - for whatever reasons that I don't understand) that I'll adapt into a middleware such that it can be used in Javascript.

So I'll be exploring the ergonomics of binding my own helpers.  Perhaps there could be a folder in Pocketbase that someone could drop their binding code such that Pocketbase Go compile would automatically pick it up?  I don't know if any of that is possible, but it will be very useful to have third party extensions that could be baked into Pocketbase so that people could use them.  Way down the road there could be a custom build web frontend.

Let's just say a single use case of being able to from Javascript serve a camera stream by only specifying a final web route and system device.  So you could just run it with `("/dev/video0", "/my-camera-feed", {"resolution": "640x480"})`.  That's nice.  Pocketbase then becomes a much more do-it-yourself home automation framework.  It just needs enough bindings to do common things.  Also bindings for toggling the RPI pins, etc.

Anyways, so that is where I'm going with this.

**Reply by dschissler** [reply]
@ganigeorgiev I see that the JSVM initialization code is not exposed for extension at the moment.  I'll wait on this until you have time to return to the project.
